plugins/parsers/csv/parser.go

```diff
@@ -0,0 +1,202 @@
+				} else {
+					headerNames[i] = headerNames[i] + name
+				}
+			}
+		}
+		p.ColumnNames = headerNames[p.SkipColumns:]
+	} else if len(p.ColumnNames) > 0 {
+		// if columns are named, just skip header rows
+		for i := 0; i < p.HeaderRowCount; i++ {
+			csvReader.Read()
+		}
+	} else if p.HeaderRowCount == 0 && len(p.ColumnNames) == 0 {
+		// if there is no header and no DataColumns, that's an error
+		return nil, fmt.Errorf("there must be a header if `csv_data_columns` is not specified")
+	}
+
+	table, err := csvReader.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+
+	metrics := make([]telegraf.Metric, 0)
+	for _, record := range table {
+		m, err := p.parseRecord(record)
+		if err != nil {
+			return metrics, err
+		}
+		metrics = append(metrics, m)
+	}
+	return metrics, nil
+}
+
+// ParseLine does not use any information in header and assumes DataColumns is set
+// it will also not skip any rows
+func (p *Parser) ParseLine(line string) (telegraf.Metric, error) {
+	r := bytes.NewReader([]byte(line))
+	csvReader, err := p.compile(r)
+	if err != nil {
+		return nil, err
+	}
+
+	// if there is nothing in DataColumns, ParseLine will fail
+	if len(p.ColumnNames) == 0 {
+		return nil, fmt.Errorf("[parsers.csv] data columns must be specified")
+	}
+
+	record, err := csvReader.Read()
+	if err != nil {
+		return nil, err
+	}
+	m, err := p.parseRecord(record)
+	if err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (p *Parser) parseRecord(record []string) (telegraf.Metric, error) {
+	recordFields := make(map[string]interface{})
+	tags := make(map[string]string)
+
+	// skip columns in record
+	record = record[p.SkipColumns:]
+	for i, fieldName := range p.ColumnNames {
+		if i < len(record) {
+			value := record[i]
+			value = strings.Trim(value, " ")
+			// attempt type conversions
+			if iValue, err := strconv.ParseInt(value, 10, 64); err == nil {
+				recordFields[fieldName] = iValue
+			} else if fValue, err := strconv.ParseFloat(value, 64); err == nil {
+				recordFields[fieldName] = fValue
+			} else if bValue, err := strconv.ParseBool(value); err == nil {
+				recordFields[fieldName] = bValue
+			} else {
+				recordFields[fieldName] = value
+			}
+		}
+	}
+
+	// add default tags
+	for k, v := range p.DefaultTags {
+		tags[k] = v
+	}
+
+	// will default to plugin name
+	measurementName := p.MetricName
+	if recordFields[p.MeasurementColumn] != nil {
+		measurementName = recordFields[p.MeasurementColumn].(string)
+	}
+
+	for _, tagName := range p.TagColumns {
+		if recordFields[tagName] == nil {
+			return nil, fmt.Errorf("could not find field: %v", tagName)
+		}
+		tags[tagName] = recordFields[tagName].(string)
+		delete(recordFields, tagName)
+	}
+
+	metricTime := time.Now()
+	if p.TimestampColumn != "" {
+		if recordFields[p.TimestampColumn] == nil {
+			return nil, fmt.Errorf("timestamp column: %v could not be found", p.TimestampColumn)
+		}
+		tStr := recordFields[p.TimestampColumn].(string)
```

Comment:
Could panic, maybe use `record` with two return value form:
```
timeColumn := record[p.TimestampColumn]
if timeColumn != nil {
    //
}

timeString, ok := col.(string); ok {
}
```

Might be easier to deal with errors if you put this into a function.